src/location/util.rs

```diff
@@ -15,51 +15,512 @@
+            let role = ObServerRole::from_int(role);
+            let replica_type = ObReplicaType::from_int(replica_type);
+            let status = ObServerStatus::from_string(status);
+            let mut observer_addr = ObServerAddr::new();
+            observer_addr.address(svr_ip);
+            observer_addr.set_sql_port(sql_port);
+            observer_addr.set_svr_port(svr_port);
+
+            let observer_info = ObServerInfo { stop_time, status };
+            if !observer_info.is_active() {
+                warn!(
+                    "ObTableLocation::get_table_location_from_remote: inactive observer found, \
+                    server_info:{:?}, addr:{:?}",
+                    observer_info, observer_addr
+                );
+                continue;
+            }
+
+            let replica = ReplicaLocation {
+                addr: observer_addr.clone(),
+                info: observer_info.clone(),
+                role: role.clone(),
+                replica_type,
+            };
+
+            if !replica.is_valid() {
+                warn!(
+                    "ObTableLocation::get_table_location_from_remote: inactive observer replica found, \
+                    server_info:{:?}, addr:{:?}",
+                    observer_info, observer_addr
+                );
+                continue;
+            }
+            // end of buildReplicaLocation in java client
+
+            let location = partition_location
+                .entry(partition_id)
+                .or_insert(ObPartitionLocation {
+                    leader: None,
+                    followers: vec![],
+                });
+
+            match role {
+                ObServerRole::Leader => location.leader = Some(replica),
+                ObServerRole::Follower => location.followers.push(replica),
+                _ => (),
+            };
+        }
+
+        // Check partition info.
+        for part_idx in 0..table_entry.partition_num {
+            // get real partition id
+            let part_id = if ob_vsn_major() >= 4 {
+                table_entry
+                    .part_tablet_id_map()
+                    .and_then(|m| m.get(&part_idx).copied())
+                    .unwrap_or(part_idx)
+            } else {
+                part_idx
+            };
+
+            let location = partition_location.get(&part_id);
+            if location.is_none() {
+                error!("Location::get_table_location_from_remote: partition num={} is not exists, table={:?}, locations={:?}",
+                   part_id, table_entry, partition_location);
+                return Err(CommonErr(
+                    CommonErrCode::PartitionError,
+                    format!("Location::get_table_location_from_remote: Table maybe dropped. partition num={part_id} is not exists, table={table_entry:?}, locations={partition_location:?}"),
+                ));
+            }
+
+            if location.unwrap().leader.is_none() {
```

Comment:
How about using `match` to handle different cases of `locaiton`?

Reply:
Fixed